src/filewatch/filewatch_inotify.c

```diff
@@ -11,59 +11,239 @@
+#define HT_KEY_PRINTABLE(key)          ((void*)(key))
+#define HT_VALUE_FMT                   "s"
+#define HT_VALUE_PRINTABLE(val)        "(empty)"
+#define HT_DECL
+#define HT_IMPL
+#include "hashtable_incproxy.inc.h"
+
 struct {
 	int inotify;
-	ht_int2ptr_ts_t wd_to_watch;
+	ht_int2wdrecord_t wd_records;
+	ht_str2ptr_t dirpath_to_dirwatch;
+	ht_filewatchset_t updated_watches;
 	SDL_mutex *modify_mtx;
-	DYNAMIC_ARRAY(FileWatch*) deactivate_list;
 } *_fw_globals;
 
+static WDRecord *wdrecord_get(int wd, bool create) {
+	WDRecord *r = NULL;
+
+	if(!ht_int2wdrecord_get_ptr_unsafe(&FW.wd_records, wd, &r, create)) {
+		if(create) {
+			*r = (WDRecord) { 0 };
+		}
+	}
+
+	return r;
+}
+
+static void wdrecord_delete(int wd) {
+	ht_int2wdrecord_unset(&FW.wd_records, wd);
+}
+
+static int wdrecord_incref(WDRecord *r) {
+	return r->refs++;
+}
+
+static int wdrecord_decref(WDRecord *r) {
+	int32_t rc = --r->refs;
+	assert(rc >= 0);
+	return rc;
+}
+
+static DirWatch *dirwatch_get(const char *path, bool create) {
+	DirWatch *dw = ht_get(&FW.dirpath_to_dirwatch, path, NULL);
+
+	if(dw || !create) {
+		if(dw) {
+			assert(dw->path != NULL);
+			assert(!strcmp(dw->path, path));
+		}
+
+		return dw;
+	}
+
+	int wd = inotify_add_watch(FW.inotify, path, WATCH_FLAGS);
+
+	if(UNLIKELY(wd == INVALID_WD)) {
+		log_error("%s: %s", path, strerror(errno));
+		return NULL;
+	}
+
+	dw = calloc(1, sizeof(*dw));
+	dw->path = strdup(path);
+	dw->wd = wd;
+
+	WDRecord *wdrec = NOT_NULL(wdrecord_get(wd, true));
+	list_push(&wdrec->dirwatch_list, dw);
+	wdrecord_incref(wdrec);
+
+	ht_set(&FW.dirpath_to_dirwatch, path, dw);
+
+	return dw;
+}
+
+static void dirwatch_delete(DirWatch *dw) {
+	assert(dw->filewatch_list == NULL);
+
+	WDRecord *wdrec = NOT_NULL(wdrecord_get(dw->wd, false));
+	list_unlink(&wdrec->dirwatch_list, dw);
+	ht_unset(&FW.dirpath_to_dirwatch, dw->path);
+	free(dw->path);
+
+	if(wdrecord_decref(wdrec) == 0) {
+		inotify_rm_watch(FW.inotify, dw->wd);
+		wdrecord_delete(dw->wd);
+	}
+
+	free(dw);
+}
+
+static void dirwatch_invalidate(DirWatch *dw) {
+	log_warn("%s: Unimplemented (FIXME)", dw->path);
+}
+
 static bool filewatch_frame_event(SDL_Event *e, void *a);
 
+attr_unused
+static void dump_events(StringBuffer *sbuf, uint events) {
+	#define DELIM() do { \
+		if(sbuf->pos != sbuf->start) { \
+			strbuf_cat(sbuf, " | "); \
+		} \
+	} while(0)
+
+	#define HANDLE(evt) do { \
+		if(events & evt) { \
+			DELIM(); \
+			strbuf_cat(sbuf, #evt); \
+			events &= ~evt; \
+		} \
+	} while(0)
+
+	HANDLE(IN_ACCESS);
+	HANDLE(IN_MODIFY);
+	HANDLE(IN_ATTRIB);
+	HANDLE(IN_CLOSE_WRITE);
+	HANDLE(IN_CLOSE_NOWRITE);
+	HANDLE(IN_OPEN);
+	HANDLE(IN_MOVED_FROM);
+	HANDLE(IN_MOVED_TO);
+	HANDLE(IN_CREATE);
+	HANDLE(IN_DELETE);
+	HANDLE(IN_DELETE_SELF);
+	HANDLE(IN_MOVE_SELF);
+
+	HANDLE(IN_UNMOUNT);
+	HANDLE(IN_Q_OVERFLOW);
+	HANDLE(IN_IGNORED);
+
+	HANDLE(IN_ONLYDIR);
+	HANDLE(IN_DONT_FOLLOW);
+	HANDLE(IN_EXCL_UNLINK);
+	HANDLE(IN_MASK_CREATE);
```

Comment:
Looks like `IN_MASK_CREATE` is not available with `libnotify-kqueue` on macOS. Might need to put a check or conditional around it to see if it exists first. (Hoping this isn't load-bearing...)

See: https://github.com/libinotify-kqueue/libinotify-kqueue/blob/ea7835fcafc3cee2a0d6c0e3c8034962c48f6afe/sys/inotify.h#L81